src/osd/scrubber/scrub_backend.cc

```diff
@@ -0,0 +1,1957 @@
+using namespace std::chrono_literals;
+using namespace std::literals;
+
+#define dout_context (m_scrubber.m_osds->cct)
+#define dout_subsys ceph_subsys_osd
+#undef dout_prefix
+
+#define dout_prefix ScrubBackend::logger_prefix(_dout, this)
+
+std::ostream& ScrubBackend::logger_prefix(std::ostream* out,
+                                          const ScrubBackend* t)
+{
+  return t->m_scrubber.gen_prefix(*out) << " b.e.: ";
+}
+
+// ////////////////////////////////////////////////////////////////////////// //
+
+// for a Primary
+ScrubBackend::ScrubBackend(PgScrubber& scrubber,
+                           PGBackend& backend,
+                           PG& pg,
+                           pg_shard_t i_am,
+                           bool repair,
+                           scrub_level_t shallow_or_deep,
+                           const std::set<pg_shard_t>& acting)
+    : m_scrubber{scrubber}
+    , m_pgbe{backend}
+    , m_pg{pg}
+    , m_pg_whoami{i_am}
+    , m_repair{repair}
+    , m_depth{shallow_or_deep}
+    , m_pg_id{scrubber.m_pg_id}
+    , m_conf{m_scrubber.get_pg_cct()->_conf}
+    , clog{m_scrubber.m_osds->clog}
+{
+  m_formatted_id = m_pg_id.calc_name_sring();
+
+  m_acting_but_me.reserve(acting.size());
+  std::copy_if(acting.begin(),
+               acting.end(),
+               std::back_inserter(m_acting_but_me),
+               [i_am](const pg_shard_t& shard) { return shard != i_am; });
+
+  m_is_replicated = m_pg.get_pool().info.is_replicated();
+  m_mode_desc =
+    (m_repair ? "repair"sv
+              : (m_depth == scrub_level_t::deep ? "deep-scrub"sv : "scrub"sv));
+}
+
+// for a Replica
+ScrubBackend::ScrubBackend(PgScrubber& scrubber,
+                           PGBackend& backend,
+                           PG& pg,
+                           pg_shard_t i_am,
+                           bool repair,
+                           scrub_level_t shallow_or_deep)
+    : m_scrubber{scrubber}
+    , m_pgbe{backend}
+    , m_pg{pg}
+    , m_pg_whoami{i_am}
+    , m_repair{repair}
+    , m_depth{shallow_or_deep}
+    , m_pg_id{scrubber.m_pg_id}
+    , m_conf{m_scrubber.get_pg_cct()->_conf}
+    , clog{m_scrubber.m_osds->clog}
+{
+  m_formatted_id = m_pg_id.calc_name_sring();
+  m_is_replicated = m_pg.get_pool().info.is_replicated();
+  m_mode_desc =
+    (m_repair ? "repair"sv
+              : (m_depth == scrub_level_t::deep ? "deep-scrub"sv : "scrub"sv));
+}
+
+void ScrubBackend::update_repair_status(bool should_repair)
+{
+  dout(15) << __func__
+           << ": repair state set to :" << (should_repair ? "true" : "false")
+           << dendl;
+  m_repair = should_repair;
+  m_mode_desc =
+    (m_repair ? "repair"sv
+              : (m_depth == scrub_level_t::deep ? "deep-scrub"sv : "scrub"sv));
+}
+
+ScrubMap* ScrubBackend::new_chunk()
+{
+  dout(15) << __func__ << dendl;
+  this_chunk.emplace(m_pg_whoami);
+  return &this_chunk->received_maps[m_pg_whoami];
+}
+
+void ScrubBackend::merge_to_master_set()
```

Comment:
Sorry for not noticing this earlier.  We're trying to avoid adding any further usages of the word 'master' to the code base. Let's use authoritative rather than master here -- it's more descriptive anyway.

Reply:
done